doc/designs/functions-for-explicitly-fetched-signature-algorithms.md

```diff
@@ -0,0 +1,189 @@
+Functions for explicitly fetched PKEY algorithms
+================================================
+
+Quick background
+----------------
+
+There are several proposed designs that end up revolving around the same
+basic need, explicitly fetched signature algorithms.  The following method
+type is affected by this document:
+
+- `EVP_SIGNATURE`
+
+Public API - Add variants of `EVP_PKEY_CTX` functionality
+---------------------------------------------------------
+
+Through OTC discussions, it's been determined that the most suitable APIs to
+touch is the collection of `EVP_PKEY_` functions.  In this case, it involves
+`EVP_PKEY_sign()`, `EVP_PKEY_verify()`, `EVP_PKEY_verify_recover()` with
+associated functions.  They can be extended to accept an explicitly fetched
+algorithm of the right type, and to be able to process infinite amounts of
+data if the fetched algorithm permits it (typically, algorithms like ED25519
+or RSA-SHA256).
+
+It must be made clear that the added functionality can not be used to
+compose an algorithm from different parts.  For example, it's not possible
+to specify a `EVP_SIGNATURE` "RSA" and combine it with a parameter that
+specifies the hash "SHA256" to get the "RSA-SHA256" functionality.  For an
+`EVP_SIGNATURE` "RSA", the input is still expected to be a digest, or some
+other input that's limited to the modulus size of the RSA pkey.
+
+### Making things less confusing with distinct function names
+
+So far, `EVP_PKEY_sign()` and friends are only expected to act as
+"primitive" functions.  This design essentially proposes an extension to
```

Comment:
I think we should not be using `primitve function` here. I think this paragraph  should try to untangle the confusion that comes from mixing up two distinct classes here:
   - API type which is a _one-shot_ vs. _streaming_ (IUF as @vdukhovni noted)
   - operation type which is either a cryptographic primitive itself (particular digest/encryption/sign/verify alg. such as AES or SHA256) or a compound operation which is made of more than one cryptographic primitive (such as digest + sign) 

also putting `primitive function` to google will get me to calculus which is even more confusing. So I would not be stealing terms from math here.

Reply:
I'm actually stealing "primitive" from [PKCS#1](https://www.rfc-editor.org/rfc/rfc8017#section-5) (`EVP_PKEY_sign` (etc) functions [actually, the provider] does give direct access to them, for RSA, under the condition that the flag `RSA_NO_PADDING` is set and that no hash function has been declared with `EVP_PKEY_CTX_set_signature_md()`, and with the [I2OSP and OSP2I](https://www.rfc-editor.org/rfc/rfc8017#section-4) primitives implied).

I'm wrong when I'm saying "only", though, since also message hashes are supported by our providers (and are properly encoded/decoded when the correct flags are set, and the signature hash is specified).

Reply:
I've expanded on this text to clarify the current functionality a bit more